cmd/migrate-lid/migrate_lid.go

```diff
@@ -240,44 +258,72 @@ func migrateIndices(ctx context.Context, logger *zap.SugaredLogger, bar *progres
 	bar.ChangeMax(len(idxPaths))
 
 	indicesStart := time.Now()
-	var count int
-	var errCount int
-	var indexTime time.Duration
-	for i, ipath := range idxPaths {
-		if ctx.Err() != nil {
-			return errCount, fmt.Errorf("index migration cancelled")
-		}
-
-		start := time.Now()
-
-		timeoutCtx, timeoutCancel := context.WithTimeout(ctx, 60*time.Second)
-		defer timeoutCancel()
-
-		indexed, err := migrateIndexWithTimeout(timeoutCtx, ipath, store, force)
-		bar.Add(1) //nolint:errcheck
-		if err != nil {
-			took := time.Since(start)
-			indexTime += took
-
-			logger.Errorw("migrate index failed", "piece cid", ipath.name, "took", took.String(), "err", err)
-
-			errCount++
-			continue
-		}
-
-		if indexed {
-			count++
-			took := time.Since(start)
-			indexTime += took
-			logger.Infow("migrated index", "piece cid", ipath.name, "processed", i+1, "total", len(idxPaths),
-				"took", took.String(), "average", (indexTime / time.Duration(count)).String())
-		} else {
-			logger.Infow("index already migrated", "piece cid", ipath.name, "processed", i+1, "total", len(idxPaths))
-		}
+	var count idxCount
+	var errCount idxCount
+	var indexTime idxTime
+	var processed idxCount
+
+	queue := make(chan idxPath, len(idxPaths))
+	for _, ipath := range idxPaths {
+		queue <- ipath
+	}
+	close(queue)
+
+	var eg errgroup.Group
+	for i := 0; i < parallel; i++ {
+		eg.Go(func() error {
+			for ctx.Err() == nil {
+				select {
+				case <-ctx.Done():
+					return ctx.Err()
+				case p, ok := <-queue:
+					if !ok {
+						// Finished adding all the queued items, exit the thread
+						return nil
+					}
+					start := time.Now()
+
+					indexed, perr := migrateIndexWithTimeout(ctx, p, store, force)
+					bar.Add(1) //nolint:errcheck
+
+					took := time.Since(start)
+					indexTime.lck.Lock()
+					indexTime.t += took
+					indexTime.lck.Unlock()
```

Comment:
nit: you can probably simplify the locking here by using atomic.AddUint64 and friends

Reply:
Done